some fix first,  we would do  "optimization for limit x, y"  and "connection pool" after

willhan123 opened the pull request:
1. fix cmake in windows
2. pre_sacn wolud cause failing to  mysqldump,   did not pre_scan when mysqldump
3. force_index push down, like select .. from t1 force index(k1) where xxx;
4. fix bug, share wolud init failed， cause dead loop

willhan123 committed:
fix cmake in windows
ldd spider.dll under right way

willhan123 committed:
limit pre_scan
pre_scan would cause remote mysql net write timeout , especial for scenes like mysqldump/insert ... select ...  , because mysqldump cost lots of time for each remote shard mysql

willhan123 committed:
"force index" push down.
We would need to specify the index sometimes, push down "force index" may help.

willhan123 committed:
fix bug
share->init would cause dead loop

willhan123 committed:
optimization for limit x,y
before, query with limit x,y would fetch x+y rows from each remote mysql, then return y rows to client
after, query with limix x,y would determine the target partition first by using select count(*) in turn , then got the result from one or more partition

-----------------------------------------
In the game scene, game server always using 'limit x,y‘ to load core data from database to game server'cache  . As a result, the optimization for 'limit x,y' is necessary in game scene.

willhan123 commented:
for optimization for limit x,y  … 
there always has some change for mysql server.
1. add two member in THD:   longlong select_offset,  longlong select_limit;
2.  modify function: select_send::send_data, select_export::send_data, select_dump::send_data, select_dumpvar::send_data 。   Add following code in front of the function body.
  if (thd->select_offset != -1 && unit->offset_limit_cnt)
  {
    DBUG_ASSERT(thd->select_offset == 0);
    unit->offset_limit_cnt = thd->select_offset;
  }

willhan123 commented:
As some modifications involve changes to the mysql service layer, may you add your own  mysql server  layer  in the warehouse ?   Additional,  it will be very convenient for me to compile, debug and communicate with you.  I will look forward to it and be grateful to you.

willhan123 committed:
SHA-1: 046846cf78e8abe97e0b1e96e353328098cdb03f

* function, connection pool
1.add parameter, spider_max_connections default 0 means disable the conncetion pool;
spider_max_connections > 0 menas enable connection pool and each remote instance can have spider_max_connections conections from spider node at most.
2. spider_conn_wait_timeout defaults 10s, control the wait time when the connection pool has  spider_max_connections connections and can not create a new connection.

the core logic of connection pool is using  'pthread_cond_timedwait' when connection pool is full,  using pthread_cond_signal when spider_free_conn_from_trx
more detail see the function spider_get_conn_from_idle_connection and spider_free_conn_from_trx

willhan123 committed:
Merge branch 'master' of https://github.com/Kentoku/Spider into kentoku-branch

Conflicts:
	spd_include.h

willhan123 committed:
fix spider_tables
change the length of table_name from 64 to 256

willhan123 committed:
fix bug when using timespec in spider_get_conn_from_idle_connection()
use last_ntime to avoide integer overflow

willhan123 committed:
fix the lenth of mysql.spider_tables.table_name

Kentoku commented on CMakeLists.txt:
```diff
@@ -57,3 +57,10 @@ IF(ORACLE_INCLUDE_DIR AND ORACLE_OCI_LIBRARY)
+  ADD_CUSTOM_COMMAND(TARGET spider
+                     POST_BUILD 
+                     COMMAND if not exist ..\\..\\sql\\lib mkdir ..\\..\\sql\\lib\\plugin
+                     COMMAND copy Debug\\ha_spider.dll ..\\..\\sql\\lib\\plugin\\ha_spider.dll)
```
Why are this changes needed?
Is this only for "Debug" build?

Kentoku commented on CMakeLists.txt:
```diff
@@ -57,3 +57,10 @@ IF(ORACLE_INCLUDE_DIR AND ORACLE_OCI_LIBRARY)
+                     POST_BUILD 
+                     COMMAND if not exist ..\\..\\sql\\lib mkdir ..\\..\\sql\\lib\\plugin
+                     COMMAND copy Debug\\ha_spider.dll ..\\..\\sql\\lib\\plugin\\ha_spider.dll)
+ENDIF()
```
Why are this changes needed?
Is this only for "Debug" build?

willhan123 replied:
yes,
we debug it in windows

Kentoku replied:
I think you can debug on windows without these changes from MariaDB 10.0. Did you try?

willhan123 replied:
I would try

willhan123 replied:
It  need to mkdir sql/lib/plugin in Mariadb 10.0, too
when I change some thing about spider, I also need to copy ha_spider.dll from storage/spider to sql/lib/plugin.

Kentoku commented:
Are all changes in this pull request under the new BSD license?

Kentoku commented on ha_spider.cc:
```diff
@@ -12871,8 +12871,17 @@ int ha_spider::check_error_mode_eof(
+  if (thd->lex && thd->lex->sql_command != SQLCOM_SELECT || // such like insert .. select ..
+    select_lex && select_lex->sql_cache == SELECT_LEX::SQL_NO_CACHE //  for mysqldump 
+    )
+  {
```
I can understand your situation, but this issue is not caused by other situation. So I'll add a parameter after merging this changes for switching this operation.
This changes are ready to merge.

willhan123 replied:
A parameter must work, too.
The issue happens when a query executes a long time after pre_scan.
In our scenes, only 'mysqldump' and 'insert into select' cause it.

Kentoku replied:
I just add parameters skip_parallel_search (table parameter) and spider_skip_parallel_search (server parameter). Please use "spider_skip_parallel_search = 3" for skipping pre_scan at INSERT SELECT and SQL_NO_CACHE. I wrote more detail about these parameters in commit comment.

Kentoku commented on spd_db_mysql.cc:
```diff
@@ -5858,6 +5858,56 @@ int spider_mysql_handler::init()
+        str->append(SPIDER_SQL_INDEX_FORCE_STR,SPIDER_SQL_INDEX_FORCE_LEN);
+        str->append(SPIDER_SQL_OPEN_PAREN_STR,SPIDER_SQL_OPEN_PAREN_LEN);
+        str->append(hint->key_name.str, hint->key_name.length);
+        str->append(SPIDER_SQL_CLOSE_PAREN_STR,SPIDER_SQL_CLOSE_PAREN_LEN);
```
Please use str->reserve() and str->q_append() for performance.

willhan123 replied:
I will fix it

Kentoku commented on spd_db_mysql.cc:
```diff
@@ -9658,6 +9708,7 @@ int spider_mysql_handler::append_from(
     str->q_append(SPIDER_SQL_FROM_STR, SPIDER_SQL_FROM_LEN);
     table_name_pos = str->length();
     append_table_name_with_adjusting(str, link_idx, sql_type);
+    append_index_hint(str, link_idx, sql_type);
```
This is danger when index name is different between Spider table and data table. So I'll add a parameter after merging this changes for switching this operation. And also you can use "idx000" table parameter for adding index hints.

willhan123 replied:
I can understand and will fix it

willhan123 replied:
I add a global parameter  'spider_index_hint_pushdown'

Kentoku commented on spd_db_oracle.cc:
```diff
@@ -5232,6 +5232,15 @@ int spider_oracle_handler::init()
+  ulong sql_type
+  ) 
+{
+  return 0;
```
Please use DBUG_ENTER() and DBUG_RETURN()

willhan123 replied:
ok

Kentoku commented on spd_db_oracle.cc:
```diff
@@ -5232,6 +5232,15 @@ int spider_oracle_handler::init()
+{
+  return 0;
+}
+
```
Please add this function to spider_handlersocket_handler class too.

willhan123 replied:
ok

Kentoku commented on spd_table.cc:
```diff
@@ -5016,9 +5016,18 @@ SPIDER_SHARE *spider_get_share(
+      {
+        fprintf(stderr, " [WARN SPIDER RESULT] "
+          "Wait share->init too long, table_name %s %s %ld\n",
+          share->table_name, share->tgt_hosts[0], share->tgt_ports[0]);
```
Please set error_num. This is a temporary fix. I should fix more deeply after merging this changes.

willhan123 replied:
you are more thorough

Kentoku commented on ha_spider.cc:
```diff
@@ -259,6 +260,7 @@ ha_spider::ha_spider(
   result_list.tmp_tables_created = FALSE;
   result_list.bgs_working = FALSE;
   result_list.direct_order_limit = FALSE;
+  result_list.direct_limit_offset = FALSE;
```
Please add it in ha_spider::reset() too.

willhan123 replied:
I am sorry and I will do that

Kentoku commented on ha_spider.cc:
```diff
@@ -7472,6 +7474,29 @@ int ha_spider::rnd_next_internal(
+      {
+        // skip this spider(partition)
+        this->trx->thd->select_offset -= table_count;
+        DBUG_RETURN(check_error_mode_eof(HA_ERR_END_OF_FILE));
```
Almost O.K. But I think you should better to make it without MariaDB changes.

willhan123 replied:
I consider whether there is a better way.
We always are confused by Mariadb's execution plan, which is not suitable in Distributed environment

Kentoku replied:
If you need a lot of time for thinking about it, I can do it after merging this changes.

willhan123 replied:
ok

willhan123 committed:
some modify after kentoku check
1. add a global parameter spider_index_hint_pushdown to control 'force index' push down
2. use q_append instead of append for performance
3. use DBUG_ENTER() and DBUG_RETURN()
4. direct_limit_offset=false in ha_spider::reset ( I had done before, not in this pull request)

Kentoku commented on spd_conn.cc:
```diff
@@ -44,6 +45,9 @@ extern ulong *spd_db_att_thread_id;
 pthread_mutex_t spider_conn_id_mutex;
+pthread_mutex_t spider_ipport_count_mutex;
+pthread_mutex_t spider_conn_i_mutexs[SPIDER_MAX_PARTITION_NUM];
+pthread_cond_t  spider_conn_i_conds[SPIDER_MAX_PARTITION_NUM];
```
Please do not use static value for it.

Kentoku commented on spd_conn.cc:
```diff
@@ -44,6 +45,9 @@ extern ulong *spd_db_att_thread_id;
 extern handlerton *spider_hton_ptr;
 extern SPIDER_DBTON spider_dbton[SPIDER_DBTON_SIZE];
 pthread_mutex_t spider_conn_id_mutex;
+pthread_mutex_t spider_ipport_count_mutex;
```
I think this mutex can be divided by conn->key_hash_value.

Kentoku commented on spd_conn.cc:
```diff
@@ -255,6 +274,17 @@ void spider_free_conn_from_trx(
+            long mutex_num=0;
+            if(ip_port_conn = (SPIDER_IP_PORT_CONN*) my_hash_search_using_hash_value(&spider_ipport_conns, conn->conn_key_hash_value, (uchar*) conn->conn_key, conn->conn_key_length))
+            {/* exists */
+              if(spider_param_max_connections())
```
If spider_max_connections is changed to 0 dynamically, all waiting threads get timeout. I need to think about this behavior.

willhan123 replied:
after spider_get_conn_from_idle_connection return NULL (may be timeout),
if(conn==NULL && spider_max_connections ==0)
then, create conn ?
do you approve this behavior ?  I this way, some threads cost 20s seconds more.

In other way, we use a thd variables to keep spider_param_max_connections.  It must involve some changes about mysql server layer.

Kentoku commented on spd_conn.cc:
```diff
@@ -4247,3 +4328,195 @@ bool spider_conn_need_open_handler(
+      fprintf(stderr, "%04d%02d%02d %02d:%02d:%02d [WARN SPIDER RESULT] "
+                      "error to spider_create_ipport_conn, next_spider_conn_mutex_id is %d\n",
+        l_time->tm_year + 1900, l_time->tm_mon + 1, l_time->tm_mday,
+        l_time->tm_hour, l_time->tm_min, l_time->tm_sec, next_spider_conn_mutex_id);
```
This limitation is not good.

Kentoku commented on spd_include.h:
```diff
@@ -1325,3 +1326,17 @@ char *spider_create_string(
+    char remote_ip_str[SPIDER_CONN_META_BUF_LEN];
+    long remote_port;
+		long ip_port_count;
+		long conn_mutex_num;
```
Please don't use TAB.

willhan123 replied:
I will pay attention

Kentoku commented on scripts/install_spider.sql:
```diff
@@ -73,7 +73,7 @@ create table if not exists mysql.spider_xa_failed_log(
 create table if not exists mysql.spider_tables(
   db_name char(64) not null default '',
-  table_name char(64) not null default '',
+  table_name char(256) not null default '',
```
Why is this 256 not 199? This change is not only for mysql.spider_tables.

Kentoku commented on spd_db_mysql.cc:
```diff
@@ -12872,6 +12886,7 @@ int spider_mysql_handler::append_order_by(
           DBUG_RETURN(HA_ERR_OUT_OF_MEM);
         str->q_append(SPIDER_SQL_COMMA_STR, SPIDER_SQL_COMMA_LEN);
       }
+*/
```
Why is this changes needed?

willhan123 replied:
my mistake, it's not needed.
I am sorry, as I do not have your latest server layer code.  I remove it for compile.

Kentoku commented:
@willhan123 Are all changes in this pull request under the new BSD license?

felixliang commented:
Hi kentoku
All our spider patches are under the new BSD license.

Felixliang 2016.10.19

发件人: Kentoku
答复: Kentoku/Spider
日期: 2016年10月19日 星期三 02:06
至: Kentoku/Spider
抄送: Felixliang, Manual
主题: Re: [Kentoku/Spider] some fix first, we would do "optimization forlimit x, y" and "connection pool" after (#18)(Internet mail)

@willhan123https://github.com/willhan123 Are all changes in this pull request under the new BSD license?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHubhttps://github.com/Kentoku/Spider/pull/18#issuecomment-254590476, or mute the threadhttps://github.com/notifications/unsubscribe-auth/ABXtITvGa0pOZgHQqxxcCxTEwGlZxk8Sks5q1QrCgaJpZM4KCVJm.

willhan123 commented on spd_db_mysql.cc:
```diff
@@ -12872,6 +12886,7 @@ int spider_mysql_handler::append_order_by(
           DBUG_RETURN(HA_ERR_OUT_OF_MEM);
         str->q_append(SPIDER_SQL_COMMA_STR, SPIDER_SQL_COMMA_LEN);
       }
+*/
```
I am sorry,

Kentoku commented:
@felixliang Thank you for your answer.

Kentoku commented:
After trying debug by Will on Windows, I'll create a branch for merging this pull request, then I'll add more changes.

felixliang commented:
## hi @Kentoku, will had checked the debug in MariaDB 10.0, here is his comment:

It need to mkdir sql/lib/plugin in Mariadb 10.0, too
when I change some thing about spider, I also need to copy ha_spider.dll from storage/spider to sql/lib/plugin.

Kentoku commented:
I'm trying to check about this. Please wait.

willhan123 commented:
![maria10](https://cloud.githubusercontent.com/assets/5733949/19633369/6863888c-99e3-11e6-9967-59d895f08f13.png)
 yeah, i can upload the picture

willhan123 committed:
fix for connection pool
use spider_max_partitions, to allocate static for spider_conn_i_mutexs

Kentoku commented:
Thanks.

Kentoku commented:
Now this changes in develop/merge_tencent_patches_20161103 branch.

willhan123 commented:
@Kentoku 
hi kentoku,
What plans do we have about 'develop/merge_tencent_patches_20161103' branch, and may I do something for you?
 
Currently, we got a spider performance problem. Do you have some good suggest ?
Query: select count(*) from tb where shard_key in ('x', 'y', 'z');
Spider would dispatch "select count(0), shard_key from tb where .... order by ...." to remote MySQL, and remote MySQL would use an inefficient index and cost lots of time.
Other way, query like 'select count(0), a from tb ' would cause 'Syntax error' in MySQL version 4.x

Kentoku commented:
Hi Will,

> What plans do we have about 'develop/merge_tencent_patches_20161103' branch, and may I do something for you?

I plan to add some changes then merge it into master before Chinese New Year. If you can create some test cases (.test file), it is helpful for it.

> Currently, we got a spider performance problem. Do you have some good suggest ?
> Query: select count(*) from tb where shard_key in ('x', 'y', 'z');
> Spider would dispatch "select count(0), shard_key from tb where .... order by ...." to remote MySQL, and remote MySQL would use an inefficient index and cost lots of time.

In this case, "shard_key" is required by executor of Spider node. What kind of INDEXs are there on data node?

> Other way, query like 'select count(0), a from tb ' would cause 'Syntax error' in MySQL version 4.x

Would you please give me more detail of syntax error?

willhan123 commented:
@Kentoku 

Thank you!

1. I wolud create some test cases. 

2. The next picture is an example, and it take a bad index
![example](https://cloud.githubusercontent.com/assets/5733949/21714613/f9cf69f8-d43a-11e6-8dfa-d36c69376aac.png)



3. In MySQL 4.x
select count(*), a from tb group by a.  This is  right syntax.
select count(*), a from tb.   This is a syntax error.    ERROR 1140: Mixing of GROUP columns (MIN(),MAX(),COUNT()...) with no GROUP columns is illegal if there is no GROUP BY clause

Kentoku commented:
@willhan123 

> 1. I wolud create some test cases.

Thanks!

> 2. The next picture is an example, and it take a bad index example

Please get explain plan on Spider node for this query. It looks choosing bad index on Spider node.

> 3. In MySQL 4.x
> select count(), a from tb group by a. This is right syntax.
> select count(), a from tb. This is a syntax error. ERROR 1140: Mixing of GROUP columns (MIN(),MAX(),COUNT()...) with no GROUP columns is illegal if there is no GROUP BY clause

Thank you for reporting it. I'll create a strict group by mode for this case after merging.

Kentoku commented:
@willhan123 @felixliang
Merging is complete. Please review my changes on develop/merge_tencent_patches_20161103 branch. If your review is OK, I'll merge this branch into master.
Sorry for delay. I'll create some tests for this patch from now.

willhan123 commented:
@Kentoku 
I am very glad to hear that, and  I very much agree with your changes. I wolud merge some changes to our work branch after tesing. 
By the way, when do you plan to pull request your master branch to MariaDB after merging this branch ?

Kentoku commented:
@willhan123 
> By the way, when do you plan to pull request your master branch to MariaDB after merging this branch ?

It's probably a month later. It's after testing, merging to Spider master, then merging to MariaDB.

Kentoku commented:
@willhan123 Now this changes were merged into Spider master.

willhan123 commented:
@Kentoku

willhan123 commented:
@Kentoku

willhan123 commented:
@Kentoku 
Many thanks. 
By the way, I had received the mail you sent to me,
and we will reply to you after some discussion.

Kentoku commented:
@willhan123
Thanks.
Now latest Spider is testing for merging into MariaDB.
http://buildbot.askmonty.org/buildbot/grid?category=main&branch=ks-10.2-spider
